src/miniscript/decode.rs

```diff
@@ -213,325 +213,348 @@ impl<Pk: MiniscriptKey, Ctx: ScriptContext> TerminalStack<Pk, Ctx> {
     }
 }
 
-/// Parse a script fragment into an `Terminal`
-#[allow(unreachable_patterns)]
-pub fn parse<Ctx: ScriptContext>(
-    tokens: &mut TokenIter,
-) -> Result<Miniscript<bitcoin::PublicKey, Ctx>, Error> {
-    let mut non_term = Vec::with_capacity(tokens.len());
-    let mut term = TerminalStack(Vec::with_capacity(tokens.len()));
+macro_rules! parse_fn {
+    ($ret_ty: ty, $fn_name: ident, $pk: ident) => {
```

Comment:
I think here we could've rather parameterized the function over `ret_ty`, like
```
￼pub fn parse<Ctx: ScriptContext, Pk: ParseableKey>(
￼    tokens: &mut TokenIter,
￼) -> Result<Miniscript<Pk, Ctx>, Error> {
```
then defined a simple trait (private to this module even) `ParseableKey` which has the `from_slice` method on it.

Reply:
Agreed, this is cleaner

Reply:
The trait can not be private because `parse` is a public interface. Not adding the seal for now, but I can add it if you think it's necessary

Reply:
Yeah, if you can seal it that'd be great. Or we can do it in a later PR, it's not a huge deal.

Reply:
Done